onnxruntime/core/codegen/mti/tensor/gather_elements.cc

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/codegen/mti/tensor/gather_elements.h"
+
+#include "core/codegen/mti/mti_tvm_utils.h"
+#include <topi/transform.h>
+
+namespace onnxruntime {
+namespace tvm_codegen {
+
+tvm::Tensor GatherElements(const tvm::Tensor& t,
+                           int64_t axis,
+                           const tvm::Tensor& indices,
+                           const std::string& name) {
+  tvm::Array<tvm::Expr> output_shape;
+  int64_t indices_rank = static_cast<int64_t>(indices->shape.size());
+  // output shape is the same as indices
+  for (int64_t i = 0; i < indices_rank; ++i)
+    output_shape.push_back(indices->shape[i]);
+
+  tvm::Expr idx_upper_bound = t->shape[axis];
+  auto l = [&](const tvm::Array<tvm::Var>& ovars) {
+    tvm::Array<tvm::Expr> ivars;
+    for (int64_t i = 0; i < indices_rank; i++) {
+      if (i == axis) {
+        tvm::Array<tvm::Expr> idx_vars;
+        for (int64_t j = 0; j < indices_rank; j++)
+          idx_vars.push_back(ovars[j]);
+        tvm::Expr idx = indices(idx_vars);
+        // when idx >= 0, we take tvm::max(..., 0), because (idx < 0) is 0
+        // when idx < 0, we take idx_upper_bound + tvm::max(...), because tvm::max(idx, 0) is 0
+        tvm::Expr real_idx = tvm::max(tvm::min(idx, idx_upper_bound - 1), 0) +
+                             (idx < 0) * (idx_upper_bound + tvm::max(idx, -idx_upper_bound));
+        // tvm idx must be of Int(32)
+        ivars.push_back(tvm::cast(tvm::Int(32), real_idx));
```

Comment:
maybe better to make it a common util for index clamping, as other ops may reuse it too (Gather?)

Reply:
Done.